nni/algorithms/compression/v2/pytorch/pruning/movement_pruner.py

```diff
@@ -0,0 +1,290 @@
+    training_epochs : int
+        The total epoch number for training the model.
+        Make sure the total `optimizer.step()` in `training_epochs` is bigger than `cool_down_beginning_step`.
+    warm_up_step : int
+        The total `optimizer.step()` number before start pruning for warm up.
+        Make sure `warm_up_step` is smaller than `cool_down_beginning_step`.
+    cool_down_beginning_step: int
+        The number of steps at which sparsity stops growing, note that the sparsity stop growing doesn't mean masks not changed.
+        The sparsity after each `optimizer.step()` is:
+        total_sparsity * (1 - (1 - (current_step - warm_up_step) / (cool_down_beginning_step - warm_up_step)) ** 3).
+    """
+    def __init__(self, model: Module, config_list: List[Dict], trainer: Callable[[Module, Optimizer, Callable], None],
+                 optimizer: Optimizer, criterion: Callable[[Tensor, Tensor], Tensor], training_epochs: int, warm_up_step: int,
+                 cool_down_beginning_step: int):
+        self.trainer = trainer
+        self.optimizer = optimizer
+        self.criterion = criterion
+        self.training_epochs = training_epochs
+        self.warm_up_step = warm_up_step
+        self.cool_down_beginning_step = cool_down_beginning_step
+        assert self.warm_up_step < self.cool_down_beginning_step, '`warm_up_step` should smaller than `cool_down_beginning_step`'
+        super().__init__(model, config_list)
+
+    def _validate_config_before_canonical(self, model: Module, config_list: List[Dict]):
+        schema_list = [deepcopy(NORMAL_SCHEMA), deepcopy(EXCLUDE_SCHEMA), deepcopy(INTERNAL_SCHEMA)]
+        schema = CompressorSchema(schema_list, model, _logger)
+        schema.validate(config_list)
+
+    def cubic_schedule(self, current_step: int):
+        if self.warm_up_step < current_step <= self.cool_down_beginning_step:
+            wrapper_dict = self.get_modules_wrapper()
+            for config in self.config_list:
+                current_sparsity = config['total_sparsity'] * (1 - (1 - (current_step - self.warm_up_step) / (self.cool_down_beginning_step - self.warm_up_step)) ** 3)
+                for op_name in config['op_names']:
+                    wrapper_dict[op_name].config['total_sparsity'] = current_sparsity
+
+    def reset_tools(self):
+        if self.metrics_calculator is None:
+            self.metrics_calculator = StraightMetricsCalculator()
+        if self.sparsity_allocator is None:
+            self.sparsity_allocator = NormalSparsityAllocator(self, continuous_mask=False)
+
+        # use Adam to update the weight_score
+        params = [{"params": [p for n, p in self.bound_model.named_parameters() if "weight_score" in n and p.requires_grad]}]
+        optimizer = Adam(params, 1e-2)
+        self.step_counter = 0
+
+        # update the masks after each optimzier step
+        def _optimizer_patch():
+            optimizer.step()
```

Comment:
no need to call optimizer.step()?

Reply:
it is needed, this optimizer only handles and updates the `weight_score` in wrapper.
`_optimizer_patch` will be called after each `patched_optimzier.step()`